### DISCUSSION TITLE
Add support for Sql Managed instance for Azure

### DISCUSSION THREAD

**Discussion by matthieupetite** [infracost/infracost#1454] [Q&A]
Hello, 

I am new to the community and I try on my side to add the support of the Azure Managed Instance object on Infracost. Unfortunately I don't find the corresponding prince in the postgre database related to the object. More precisely I didn't find the corresponding price. In terraform we have the ability to create a Gen4 and Gen5 database and I don't see price for reserved instance and the licence option.

Atfer an update of the postgres database today, here is the product list 

                                   productname
----------------------------------------------------------------------------------
 SQL Managed Instance Azure Hybrid Benefit - Enterprise Edition
 SQL Managed Instance Azure Hybrid Benefit - Standard Edition
 SQL Managed Instance Business Critical - Compute Gen5
 SQL Managed Instance Business Critical - Premium Series Compute
 SQL Managed Instance Business Critical - Premium Series Memory Optimized Compute
 SQL Managed Instance Business Critical - SQL License
 SQL Managed Instance Business Critical - Storage
 SQL Managed Instance General Purpose - Compute Gen5
 SQL Managed Instance General Purpose - Premium Series Compute
 SQL Managed Instance General Purpose - Premium Series Memory Optimized Compute
 SQL Managed Instance General Purpose - SQL License
 SQL Managed Instance General Purpose - Storage
 SQL Managed Instance Hyperscale - Compute Gen4
 SQL Managed Instance Hyperscale - Compute Gen5
 SQL Managed Instance Hyperscale - SQL License
 SQL Managed Instance Hyperscale - Storage
 SQL Managed Instance - LTR Backup Storage
 SQL Managed Instance PITR Backup Storage

In that case I can only do the pricing for gen5 Business Critical and Gen Purpose without the ability to used the reserved instance option and the licence discound which will provide only Part of the solution.

Did you have any advice on that ?

How can I still contribute ?

Regards 

Matthieu

**Comment by aliscott** [comment]
Thanks @matthieupetite, I'll have a look at the Managed SQL pricing today to get myself up to speed with it and see what I can find to help.

**Reply by matthieupetite** [reply]
@aliscott thanks a lot for you support,  can you please tell me how is  generated the csv file containing the price list ?

**Reply by aliscott** [reply]
@matthieupetite Here's what I've found so far:

**Gen 4**

I also can't find any prices in the Azure APIs for the Gen 4 instances, so I'm not sure they publish them here any more. I'm not sure how common it is to use these old instances, so maybe it's okay to leave it out for now?

**Reserved capacity pricing**

For reserved capacity, we don't support this for any Azure types yet. Perhaps we can also leave this out of the initial support for this resource, what do you think?

**Hybrid Benefit Pricing**

I'm seeing data in the database for the `productName = 'SQL Managed Instance General Purpose - SQL License'`. Here's my query:

```
SELECT "service", "productFamily",  "attributes" ->> 'meterName', "attributes" ->> 'skuName', "prices"
FROM products
WHERE "vendorName" = 'azure'
  AND "region" = 'Global'
  AND "attributes" ->> 'productName' = 'SQL Managed Instance General Purpose - SQL License';
```

If we take the price returned by this and multiply it by the number of cores of the instance we get the Hybrid Benefit discount.
For the `azurerm_sql_database` resource we have a [SQL License cost component](https://github.com/infracost/infracost/blob/master/internal/resources/azure/sql_database.go#L267) that we show that has this price, unless the resource is using Hybrid Benefit pricing which we detect by checking the `license_type` attribute on the Terraform resource. I think a similar approach could work for this resource.

> can you please tell me how is generated the csv file containing the price list ?

We do this in our [Cloud Pricing API](https://github.com/infracost/cloud-pricing-api). You can run it locally and run `npm run data:scrape --only=azure:retail` to download all the Azure JSON files, but this can take several hours. The pricing dump you have should have all the up-to-date pricing data anyway.

**Reply by matthieupetite** [reply]
Thanks @aliscott for all those information, based on that I will leave the Gen 4 out and work as you said to handle the licence issue. 

Regards

Matthieu

**Reply by aliscott** [reply]
🙌 great thanks for picking this up! Please reach out if I can help more, we're also on the [Infracost Slack](https://infracost.io/community-chat) if that helps.
